node/actors/network/src/gossip/mod.rs

```diff
@@ -103,4 +117,72 @@ impl Network {
         })
         .await;
     }
+
+    /// Task that keeps hearing about new votes and updates the L1 batch qc.
+    /// It will propagate the QC if there's enough votes.
+    pub(crate) async fn update_batch_qc(&self, ctx: &ctx::Ctx) -> anyhow::Result<()> {
+        // TODO This is not a good way to do this, we shouldn't be verifying the QC every time
+        // Can we get only the latest votes?
+        let attesters = self.genesis().attesters.as_ref().context("attesters")?;
+        loop {
+            let mut sub = self.batch_votes.subscribe();
+            let votes = sync::changed(ctx, &mut sub)
+                .await
+                .context("batch votes")?
+                .clone();
+
+            // Check next QC to collect votes for.
+            let new_qc = self
+                .last_viewed_qc
+                .clone()
+                .map(|qc| {
+                    attester::BatchQC::new(
+                        attester::Batch {
+                            number: qc.message.number.next(),
+                        },
+                        self.genesis(),
+                    )
+                })
+                .unwrap_or_else(|| {
+                    attester::BatchQC::new(
+                        attester::Batch {
+                            number: attester::BatchNumber(0),
+                        },
+                        self.genesis(),
+                    )
+                })
+                .context("new qc")?;
+
+            // Check votes for the correct QC.
+            for (_, sig) in votes.0 {
+                if self
+                    .batch_qc
+                    .clone()
```

Comment:
It's not obvious to me why we clone the map in each step. Don't we want to accumulate all signatories before the `weight` is checked next? 

Also, shouldn't these signatures added stay in `batch_qc` after `update_batch_qc` returns? Why doesn't it take `&mut self`? 

If this is just done for the sake of checking individual voters, why does the total weight get checked here as well?

Reply:
Right, thanks. To be clear, I didn't start looking at this PR to understand the codebase, rather, I asked this question because I really didn't understand what the code in the PR was trying to achieve, and the description didn't contain any caveats about it being incomplete or nonsensical.